experimental/adaptive-dialog/docs/recognizers-rules-steps-reference.md

```diff
@@ -0,0 +1,739 @@
+// a number from the user.
+getNumberDialog.AddRule(new EventRule(["Contoso.travelBot.getNumber"], 
+    steps: new List<IDialog> {
+        new NumberInput<int>()
+        {
+            Property = "turn.contoso.travelBot.numberPrompt",
+            Prompt = new ActivityTemplate("{turn.contoso.travelBot.numberMessage}")
+        }
+    });
+```
+
+## Steps
+_Steps_ help put together the flow of conversation when a specific event is captured via a Rule. **_Note:_** unlike Waterfall dialog where each step is a function, each step in an Adaptive dialog is in itself a dialog. This enables adaptive dialogs by design to 
+- have a much cleaner ability to handle and deal with interruptions.
+- branch conditionally based on context or state.
+
+Adaptive dialogs support the following steps - 
+- Sending a response
+    - [SendActivity](#SendActivity)
+- Memory manipulation
+    - [SaveEntity](#SaveEntity)
+    - [EditArray](#EditArray)
+    - [InitProperty](#InitProperty)
+    - [SetProperty](#SetProperty)
+    - [DeleteProperty](#DeleteProperty)
+- Conversational flow and dialog management
+    - [IfCondition](#IfCondition)
+    - [SwitchCondition](#SwitchCondition)
+    - [EndTurn](#EndTurn)
+    - [BeginDialog](#BeginDialog)
+    - [EndDialog](#EndDialog)
+    - [CancelAllDialog](#CancelAllDialog)
+    - [ReplaceDialog](#ReplaceDialog)
+    - [RepeatDialog](#RepeatDialog)
+    - [EmitEvent](#EmitEvent)
+- Roll your own code
+    - [CodeStep](#CodeStep)
+    - [HttpRequest](#HttpRequest)
+- Tracing and logging
+    - [TraceActivity](#TraceActivity)
+    - [LogStep](#LogStep)
+
+### SendActivity
+Used to send an activity to user. 
+
+``` C#
+// Example of a simple SendActivity step
+var greetUserDialog = new AdaptiveDialog("greetUserDialog");
+greetUserDialog.AddRule(new IntentRule("greetUser", 
+    steps: new List<IDialog>() {
+        new SendActivity("Hello")
+}));
+
+// Example that includees reference to property on bot state.
+var greetUserDialog = new AdaptiveDialog("greetUserDialog");
+greetUserDialog.AddRule(new IntentRule("greetUser", 
+    steps: new List<IDialog>() {
+        new TextInput("user.name", "What is your name?"),
+        new SendActivity("Hello, {user.name}")
+}));
+```
+See [here][3] to learn more about using language generation instead of hard coding actual response text in SendActivity.
+
+### SaveEntity
+Use this step to promte an entity (from a recognizer) into a different memory scope. By default entities from recognizer are available under the [turn scope][4] and the life time of all information under that scope is the end of that turn of conversation. 
+
+``` C#
+var greetUserDialog = new AdaptiveDialog("greetUserDialog");
+greetUserDialog.AddRule(new IntentRule("greetUser", 
+    steps: new List<IDialog>() {
+        // Save the userName entitiy from a recognizer.
+        new SaveEntity("user.name", "turn.entities.userName[0]"),
```

Comment:
it will be good to link to one of the actual samples, to explain were does userName[0] comes from